---
    title: "1. analysis preprocess en4"
    linkTitle: "1. analysis preprocess en4"
    weight: 5

    description: >
        1. analysis preprocess en4 example.
---
Script for processing raw EN4 data prior to analysis.
See docstring of `Profile.process_en4()` for more specifics on what it does.

This script will just load modules, read in raw EN4 data, cut out a geographical box, call the processing routine and write the processed data to file.

You don't have to do this for each EN4 file individually if you don't want, you can read in multiple using `multiple = True` on the creation of the profile object. However, if analysing model data in parallel chunks, you may want to split up the processing into smaller files to make the analysis faster.


```python
import sys

# IF USING A DEVELOPMENT BRANCH OF COAST, ADD THE REPOSITORY TO PATH:
# sys.path.append('<PATH_TO_COAST_REPO')
import coast
import pandas as pd
from datetime import datetime
from os import path

print("Modules loaded")

# File paths - input en4, output processed file and read config file
"""
fn_prof = "<PATH_TO_RAW_EN4_DATA_FILE(S)>"
fn_out = "<PATH_TO_OUTPUT_LOCATION_FOR_PROCESSED_PROFILES>"
fn_cfg_prof = "<PATH_TO_COAST_PROFILE_CONFIG_FILE>"
"""

fn_out = "./output.nc"
fn_prof = path.join('./example_files', "coast_example_en4_201008.nc")
fn_prof = path.join('/Users/jeff/Downloads/EN.4.2.2.profiles.g10.2022', "EN.4.2.2.f.profiles.g10.2022*.nc")
fn_cfg_prof = path.join('/Users/jeff/gitHub/COAsT/config', "example_en4_profiles.json")


# Some important settings, easier to get at here
longitude_bounds = [-15, 15]  # Geo box to cut out from data (match to model)
latitude_bounds = [40, 65]
multiple = True  # Reading multple files?
```

    Modules loaded



```python
fn_prof = "/Users/jeff/GitHub/COAsT/example_files/coast_example_en4_201008.nc"
```


```python
# Some important settings, easier to get at here
longitude_bounds = [0, 360]  # Geo box to cut out from data (match to model)
latitude_bounds = [-15, 15]
multiple = True  # Reading multple files?
```

Create profile object containing data


```python
profile = coast.Profile(config=fn_cfg_prof)
profile.read_en4(fn_prof, multiple=multiple)
```

    /Users/jeff/gitHub/COAsT/config/example_en4_profiles.json



    ---------------------------------------------------------------------------

    FileNotFoundError                         Traceback (most recent call last)

    /tmp/ipykernel_3793/3838477898.py in <cell line: 1>()
    ----> 1 profile = coast.Profile(config=fn_cfg_prof)
          2 profile.read_en4(fn_prof, multiple=multiple)


    /usr/share/miniconda/envs/coast/lib/python3.8/site-packages/coast/data/profile.py in __init__(self, dataset, config)
         62         debug(f"Creating a new {get_slug(self)}")
         63         self.config = config
    ---> 64         super().__init__(self.config)
         65 
         66         # If dataset is provided, put inside this object


    /usr/share/miniconda/envs/coast/lib/python3.8/site-packages/coast/data/index.py in __init__(self, config)
         32         if config:
         33             print(config)
    ---> 34             self.json_config = ConfigParser(config)
         35             self.chunks = self.json_config.config.chunks
         36             self.dim_mapping = self.json_config.config.dataset.dimension_map


    /usr/share/miniconda/envs/coast/lib/python3.8/site-packages/coast/data/config_parser.py in __init__(self, json_path)
         16             json_path (Union[Path, str]): path to json config file.
         17         """
    ---> 18         with open(json_path, "r") as j:
         19             json_content = json.loads(j.read())
         20             conf_type = ConfigTypes(json_content[ConfigKeys.TYPE])


    FileNotFoundError: [Errno 2] No such file or directory: '/Users/jeff/gitHub/COAsT/config/example_en4_profiles.json'



```python
profile.dataset
```


    ---------------------------------------------------------------------------

    NameError                                 Traceback (most recent call last)

    /tmp/ipykernel_3793/2681786031.py in <cell line: 1>()
    ----> 1 profile.dataset
    

    NameError: name 'profile' is not defined


Get geographical indices to extract.


```python
profile = profile.subset_indices_lonlat_box(longitude_bounds, latitude_bounds)
```


    ---------------------------------------------------------------------------

    NameError                                 Traceback (most recent call last)

    /tmp/ipykernel_3793/2450699116.py in <cell line: 1>()
    ----> 1 profile = profile.subset_indices_lonlat_box(longitude_bounds, latitude_bounds)
    

    NameError: name 'profile' is not defined



```python
profile.quick_plot()
```


    ---------------------------------------------------------------------------

    NameError                                 Traceback (most recent call last)

    /tmp/ipykernel_3793/1711720463.py in <cell line: 1>()
    ----> 1 profile.quick_plot()
    

    NameError: name 'profile' is not defined


Cut out a time slice of the data.


```python
profile = profile.time_slice(date0=datetime(2010, 1, 1), date1=datetime(2010, 1, 20))
```


    ---------------------------------------------------------------------------

    NameError                                 Traceback (most recent call last)

    /tmp/ipykernel_3793/567752478.py in <cell line: 1>()
    ----> 1 profile = profile.time_slice(date0=datetime(2010, 1, 1), date1=datetime(2010, 1, 20))
    

    NameError: name 'profile' is not defined


Process the extracted data into new processed profile.


```python
processed_profile = profile.process_en4()
```


    ---------------------------------------------------------------------------

    NameError                                 Traceback (most recent call last)

    /tmp/ipykernel_3793/2338363098.py in <cell line: 1>()
    ----> 1 processed_profile = profile.process_en4()
    

    NameError: name 'profile' is not defined


Sometimes the following line is needed to avoid an error::  
`processed_profile.dataset["time"] = ("id_dim", pd.to_datetime(processed_profile.dataset.time.values))`

Write processed profiles to file.


```python
processed_profile.dataset.to_netcdf(fn_out)
```


    ---------------------------------------------------------------------------

    NameError                                 Traceback (most recent call last)

    /tmp/ipykernel_3793/1086284925.py in <cell line: 1>()
    ----> 1 processed_profile.dataset.to_netcdf(fn_out)
    

    NameError: name 'processed_profile' is not defined



```python

```
